src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -231,4 +225,95 @@ protected void performCleanup() {
+     */
+    static final class PhantomCleanableList {
+        private static final int MIN_CAPACITY = 16;
+        private final Object lock = new Object();
+        private PhantomCleanable<?>[] arr;
+        private int size;
+
+        public PhantomCleanableList() {
+            this.arr = new PhantomCleanable<?>[MIN_CAPACITY];
+            this.size = 0;
+        }
+
+        /**
+         * Returns true if cleanable list is empty.
+         *
+         * @return true if the list is empty
+         */
+        public boolean isEmpty() {
+            synchronized (lock) {
+                return size == 0;
+            }
+        }
+
+        /**
+         * Insert this PhantomCleanable in the list.
+         */
+        public void insert(PhantomCleanable<?> phc) {
+            synchronized (lock) {
+                // Resize if needed.
+                int oldLen = arr.length;
+                if (oldLen <= size) {
+                    int newLen = ArraysSupport.newLength(oldLen, 1, oldLen);
+                    arr = Arrays.copyOf(arr, newLen);
+                }
+                // Inserting at the end, record the indexes.
+                phc.index = size;
+                arr[size] = phc;
+                size++;
+            }
+        }
+
+        /**
+         * Remove this PhantomCleanable from the list.
+         *
+         * @return true if Cleanable was removed or false if not because
+         * it had already been removed before
+         */
+        public boolean remove(PhantomCleanable<?> phc) {
+            synchronized (lock) {
+                int thisIdx = phc.index;
+                if (thisIdx == -1) {
+                    // Not in the list.
+                    return false;
+                }
+
+                // Unlink PhantomCleanable.
+                assert arr[phc.index] == phc;
+                phc.index = -1;
+
+                int lastIdx = size - 1;
+                if (lastIdx != thisIdx) {
+                    // Move the last, still alive element at current index,
+                    // overwriting the removed one. Update its index to a new location.
+                    PhantomCleanable<?> last = arr[lastIdx];
+                    last.index = thisIdx;
+                    arr[thisIdx] = last;
+                }
+
+                // Cut the tail.
+                arr[lastIdx] = null;
+                size--;
+
+                // Capacity control: trim the backing storage if it looks like
+                // we have a lot of wasted space there. Resizing on insertion would
+                // double the array size, so this is our best case. Therefore, we want
+                // to check if less than a quarter of the array is busy. We also do not
+                // want to cause an immediate resize on next insertion.
+                if ((size < arr.length / 4) && (size > MIN_CAPACITY)) {
+                    int newLen = ArraysSupport.newLength(size, 1, size);
```

Comment:
This code isn't immediately clear that it's trying to half the array size, as you are using the size computation method for array growth.  Can you add a remark for so?

Reply:
None of this is actually needed once we do smarter thing than relying on array resizes (and copies). I'll push new implementation soon.